File: pkg/go/transformer/mod-to-json.go
package transformer

import (
	"fmt"
	"strings"

	"github.com/hashicorp/go-multierror"
	"gopkg.in/yaml.v3"
)

type ModFile struct {
	Schema   string   `json:"schema"`
	Module   string   `json:"module"`
	Contents []string `json:"contents"`
}

type YAMLModFile struct {
	Schema   yaml.Node `yaml:"schema"`
	Module   yaml.Node `yaml:"module"`
	Contents yaml.Node `yaml:"contents"`
}

type ModFileValidationErrorMetadata struct{}

// ModFileValidationError is an error occurred during validation of the mod.fga file. Line and
// column number provided are one based.
type ModFileValidationError struct {
	Line, Column int
	Msg          string
}

func (e *ModFileValidationError) Error() string {
	return fmt.Sprintf("validation error at line=%d, column=%d: %s", e.Line, e.Column, e.Msg)
}

type ModFileValidationMultipleError multierror.Error

func (e *ModFileValidationMultipleError) Error() string {
	errors := e.Errors

	pluralS := ""
	if len(errors) > 1 {
		pluralS = "s"
	}

	errorsString := []string{}
	for _, item := range errors {
		errorsString = append(errorsString, item.Error())
	}

	return fmt.Sprintf("%d error%s occurred:\n\t* %s\n\n", len(errors), pluralS, strings.Join(errorsString, "\n\t* "))
}

const (
	stringNode = "!!str"
	seqNode    = "!!seq"
)

// TransformModFile transforms a mod.fga and validates the fields are correct.
func TransformModFile(data string) (*ModFile, error) { //nolint:cyclop
	yamlModFile := &YAMLModFile{}

	err := yaml.Unmarshal([]byte(data), yamlModFile)
	if err != nil {
		return nil, err
	}

	modFile := &ModFile{}
	errors := &multierror.Error{}

	switch {
	case yamlModFile.Schema.IsZero():
		errors = multierror.Append(errors, &ModFileValidationError{
			Msg:    "missing schema field",
			Line:   1,
			Column: 1,
		})
	case yamlModFile.Schema.Tag != stringNode:
		errors = multierror.Append(errors, &ModFileValidationError{
			Msg:    "unexpected schema type, expected string got value " + yamlModFile.Schema.Value,
			Line:   yamlModFile.Schema.Line,
			Column: yamlModFile.Schema.Column,
		})
	case yamlModFile.Schema.Value != "1.2":
		errors = multierror.Append(errors, &ModFileValidationError{
			Msg:    "unsupported schema version, fga.mod only supported in version `1.2`",
			Line:   yamlModFile.Schema.Line,
			Column: yamlModFile.Schema.Column,
		})
	default:
		modFile.Schema = yamlModFile.Schema.Value
	}

	switch {
	case yamlModFile.Module.IsZero():
		errors = multierror.Append(errors, &ModFileValidationError{
			Msg:    "missing module field",
			Line:   1,
			Column: 1,
		})
	case yamlModFile.Module.Tag != stringNode:
		errors = multierror.Append(errors, &ModFileValidationError{
			Msg:    "unexpected module type, expected string got value " + yamlModFile.Module.Value,
			Line:   yamlModFile.Module.Line,
			Column: yamlModFile.Module.Column,
		})
	default:
		modFile.Module = yamlModFile.Module.Value
	}

	switch {
	case yamlModFile.Contents.IsZero():
		errors = multierror.Append(errors, &ModFileValidationError{
			Msg:    "missing contents field",
			Line:   1,
			Column: 1,
		})
	case yamlModFile.Contents.Tag != seqNode:
		errors = multierror.Append(errors, &ModFileValidationError{
			Msg:    "unexpected contents type, expected list of strings got value " + yamlModFile.Contents.Value,
			Line:   yamlModFile.Contents.Line,
			Column: yamlModFile.Contents.Column,
		})
	default:
		contents := []string{}
		for _, file := range yamlModFile.Contents.Content {
			contents = append(contents, file.Value)

			if file.Tag != stringNode {
				errors = multierror.Append(errors, &ModFileValidationError{
					Msg:    "unexpected contents item type, expected string got value " + file.Value,
					Line:   file.Line,
					Column: file.Column,
				})
			} else if !strings.HasSuffix(file.Value, ".fga") {
				errors = multierror.Append(errors, &ModFileValidationError{
					Msg:    "contents items should use fga file extension, got " + file.Value,
					Line:   file.Line,
					Column: file.Column,
				})
			}
		}

		modFile.Contents = contents
	}

	if len(errors.Errors) != 0 {
		return nil, &ModFileValidationMultipleError{
			Errors: errors.Errors,
		}
	}

	return modFile, nil
}

File: pkg/js/transformer/modules/mod-to-json.ts
import { parseDocument, LineCounter, Scalar, isSeq, isNode, YAMLSeq } from "yaml";
import { LinePos } from "yaml/dist/errors";
import { FGAModFileValidationError, FGAModFileValidationSingleError } from "../../errors";

/**
 * Represents a property contained in the `fga.mod` file, includes the value as well as line an
 * column information to allow improved error reporting.
 */
export interface ModFileProperty<T> {
  /**
   * The value of the property.
   */
  value: T

  /**
   * The start and end line number of the property.
   */
  line: {
    start: number;
    end: number;
  }

  /**
   * The start and end column number of the property.
   */
  column: {
    start: number;
    end: number;
  }
}

/**
 * An `fga.mod` file represented as JSON
 */
export interface ModFile {
  /**
   * The schema version. This currently will only be 1.2
   */
  schema: ModFileProperty<string>
  /**
   * The module name.
   */
  module: ModFileProperty<string>
  /**
   * The individual files that make up the modular model.
   */
  contents: ModFileProperty<ModFileProperty<string>[]>
}

/**
 * Gets the line and column data from the `linePos` returned from parsing a yaml file.
 * @param linePos - The `linePos` property.
 * @returns Line and column data.
 */
function getLineAndColumnFromLinePos(
  linePos?: [LinePos] | [LinePos, LinePos]
): { line: { start: number; end: number; }, column: { start: number; end: number; }} {
  if (linePos === undefined) {
    return { line: { start: 1, end: 1 }, column: { start: 1, end: 1 } };
  }
  // eslint-disable-next-line prefer-const
  let [start, end] = linePos;
  if (end === undefined) {
    end = start;
  }
  return {
    line: {
      start: start.line ,
      end: end.line
    },
    column: {
      start: start.col,
      end: end.col
    }
  };
}

/**
 * Gets the line and column data for a node in a yaml doc.
 * @param node - The node from the yaml doc.
 * @param counter - The instance of LineCounter that was passed to `parseDocument`.
 * @returns Line and column data.
 */
function getLineAndColumnFromNode(
  node: unknown,
  counter: LineCounter
): { line: { start: number; end: number; }, column: { start: number; end: number; }} {
  if (!isNode(node) || !node.range) {
    return { line: { start: 1, end: 1 }, column: { start: 1, end: 1 } };
  }

  const start = counter.linePos(node.range[0]);
  const end = counter.linePos(node.range[1]);
  return {
    line: {
      start: start.line,
      end: end.line
    },
    column: {
      start: start.col ,
      end: end.col
    }
  };
}

/**
 * Transforms an `fga.mod` file into the JSON representation and validate the fields are correct.
 * @param {string} modFile - The `fga.mod` file
 * @returns {ModFile} The jSON representation of the `fga.mod` file.
 */
export const transformModFileToJSON = (modFile: string): ModFile => {
  const lineCounter = new LineCounter();
  const yamlDoc = parseDocument(modFile, {
    lineCounter,
    keepSourceTokens: true
  });

  const errors: FGAModFileValidationSingleError[] = [];
  // Copy over any general yaml parsing errors
  for (const error of yamlDoc.errors) {
    errors.push(new FGAModFileValidationSingleError({
      msg: error.message,
      ...getLineAndColumnFromLinePos(error.linePos)
    }));
  }

  const parsedModFile: Partial<ModFile> = {};

  const schemaNode = yamlDoc.get("schema", true) as Scalar<string>;
  if (!schemaNode) {
    errors.push(new FGAModFileValidationSingleError({
      msg: "missing schema field",
      ...getLineAndColumnFromLinePos()
    }));
  } else if (typeof schemaNode.value !== "string") {
    errors.push(new FGAModFileValidationSingleError({
      msg: `unexpected schema type, expected string got value ${schemaNode.value}`,
      ...getLineAndColumnFromNode(schemaNode, lineCounter)
    }));
  } else if (schemaNode.value !== "1.2") {
    errors.push(new FGAModFileValidationSingleError({
      msg: "unsupported schema version, fga.mod only supported in version `1.2`",
      ...getLineAndColumnFromNode(schemaNode, lineCounter)
    }));
  } else {
    parsedModFile.schema = {
      value: schemaNode.value,
      ...getLineAndColumnFromNode(schemaNode, lineCounter)
    };
  }

  const moduleNode = yamlDoc.get("module", true) as Scalar<string>;
  if (!moduleNode) {
    errors.push(new FGAModFileValidationSingleError({
      msg: "missing module field",
      ...getLineAndColumnFromLinePos()
    }));
  } else if (typeof moduleNode.value !== "string") {
    errors.push(new FGAModFileValidationSingleError({
      msg: `unexpected module type, expected string got value ${moduleNode.value}`,
      ...getLineAndColumnFromNode(moduleNode, lineCounter)
    }));
  } else {
    parsedModFile.module = {
      value: moduleNode.value,
      ...getLineAndColumnFromNode(moduleNode, lineCounter)
    };
  }

  const contentsNode = yamlDoc.get("contents", true) as YAMLSeq<Scalar>;
  if (!contentsNode) {
    errors.push(new FGAModFileValidationSingleError({
      msg: "missing contents field",
      ...getLineAndColumnFromLinePos()
    }));
  } else if (!isSeq(contentsNode)) {
    const node = yamlDoc.get("contents", true);
    const contents = yamlDoc.get("contents");
    errors.push(new FGAModFileValidationSingleError({
      msg: `unexpected contents type, expected list of strings got value ${contents}`,
      ...getLineAndColumnFromNode(node, lineCounter)
    }));
  } else {
    const contents = yamlDoc.get("contents") as YAMLSeq<Scalar>;
    const contentsValue = [];
    for (const file of contents.items) {
      if (typeof file.value !== "string") {
        errors.push(new FGAModFileValidationSingleError({
          msg: `unexpected contents item type, expected string got value ${file.value}`,
          ...getLineAndColumnFromNode(file, lineCounter)
        }));
        continue;
      }

      if (!file.value.endsWith(".fga")) {
        errors.push(new FGAModFileValidationSingleError({
          msg: `contents items should use fga file extension, got ${file.value}`,
          ...getLineAndColumnFromNode(file, lineCounter)
        }));
        continue;
      }

      contentsValue.push({
        value: file.value,
        ...getLineAndColumnFromNode(file, lineCounter)
      });
    }
    const node = yamlDoc.get("contents", true);
    parsedModFile.contents = {
      value: contentsValue,
      ...getLineAndColumnFromNode(node, lineCounter)
    };

  }

  if (errors.length) {
    throw new FGAModFileValidationError(errors);
  }

  return parsedModFile as ModFile;
};
